vm_insnhelper.c

```diff
@@ -3872,6 +3878,67 @@ vm_opt_regexpmatch2(VALUE recv, VALUE obj)
     }
 }
 
+#define PC_OFFSET(insn, ci, cc) leaf ? 0 : attr_width_##insn(ci, cc)
+
+static void
+vm_change_insn(rb_control_frame_t *cfp, int pc_offset, int insn)
+{
+    *((VALUE *)cfp->pc - pc_offset) = (VALUE)rb_vm_insn_insn2addr(insn, ruby_vm_event_enabled_flags & ISEQ_TRACE_EVENTS);
+}
+
+static VALUE
+vm_specialize_insn(rb_control_frame_t *cfp, int pc_offset, const struct rb_call_info *ci)
+{
+#define SP_INSN0(opt) \
+    recv = *(cfp->sp - 1); \
+    if ((val = vm_opt_##opt(recv)) != Qundef) { \
+        vm_change_insn(cfp, pc_offset, BIN(opt_##opt)); \
+        cfp->sp -= 1; \
```

Comment:
Isn't stack pointer already decremented appropriately in the insn prelude of opt_send_without_block?  Maybe I'm misunderstanding something.

Reply:
I guess you meant here https://github.com/ruby/ruby/blob/0bceb471e8d07bdc5a32de0fb67373e4e3c3d385/tool/ruby_vm/views/_insn_entry.erb#L35-L37 by "prelude of opt_send_withou_block"?

In my local insns_info.inc, `attr_popn_opt_send_without_block` is always returning 0 (since [pops exclude "..."](https://github.com/ruby/ruby/blob/0bceb471e8d07bdc5a32de0fb67373e4e3c3d385/tool/ruby_vm/models/bare_instructions.rb#L28) and opt_send_without_block's operand is "..."). So stack pointer is not decremented there in my understanding.

It seems to be usually decremented here https://github.com/ruby/ruby/blob/0bceb471e8d07bdc5a32de0fb67373e4e3c3d385/vm_insnhelper.c#L1927 and the line is not used when `SP_INSN0` is used because opt_send_without_block bypasses to run `CALL_METHOD` in that case (returning non-`Qundef`).

Reply:
Sounds good. Actually that needed to be https://github.com/ruby/ruby/pull/1970/commits/57ab71f9d5db59d258c781b75496bbd3ae0df892